File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        // 1st
        // 0 -> 1 -> 2 -> 3 -> null
        // c    t
        // 0 -> 1 -> 2 -> 3 -> null
        // n    c
        // 0 -> 1 -> 2 -> 3 -> null
        // p    c

        // 2nd
        // 0 -> 1 -> 2 -> 3 -> null
        // c    t
        // 0 -> 1 -> 2 -> 3 -> null
        // n    c
        // 0 -> 1 -> 2 -> 3 -> null
        // p    c
        ListNode curr = head;
        ListNode prev = null;

        while (curr != null) {
            ListNode temp = curr.next;
            curr.next = prev; // points to null now
            prev = curr;
            curr = temp;
        }
        return prev;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        
        # Brute force
        arr = []
        current = head

        # dump list elements into array
        while current:
            arr.append(current.val)
            current = current.next

        n = len(arr)
        for i in range(n // k):
            start = i * k
            end = start + (k-1)
            arr[start:end+1:] = arr[start:end+1:][::-1]

        dummy = ListNode(0, None)
        mover = dummy
        for i in range(len(arr)):
            mover.next = ListNode(arr[i])
            mover = mover.next
        
        return dummy.next